### PR TITLE
Add support for post_parse_analyze to PgHooks

### PR DESCRIPTION
**Pull request by jteplitz**

### CODE REVIEW

File: pgx/src/hooks.rs
Comment (workingjubilee) [inline pgx/src/hooks.rs:17]: This can be externally constructed and is thus not really "opaque".
```suggestion
// This zero-sized type is here to provide an inner type for
```

**Review by workingjubilee — COMMENTED** [review]
There's a lot of code duplication in this PR to satisfy the `cfg` requirements? But it looks almost exactly the same? Why not just `#[cfg]` on a single `let` binding inside the `pgx_post_parse_analyze` function?

**Comment by jteplitz** [comment]
> Why not just `#[cfg]` on a single `let` binding inside the `pgx_post_parse_analyze` function?

I'd like to remove the duplicate code, but am not sure how to do it. I'm still new to rust so please let me know if there's something I can learn here!

The function signature is different in pg14 since it takes an extra argument. Is there a way to conditionally compile part of the parameter list in rust? We'd need something equivalent to this C code:
```c
void pgx_post_parse_analyze(
ParseState *ps,
Query *query,
#if PG14 
JumbleState *jumble_state
#endif /* PG14 */
) {
```
My naive attempt to mark the parameter itself with a cfg attribute doesn't work
```rust
  unsafe extern "C" fn pgx_post_parse_analyze(
      parse_state: *mut pg_sys::ParseState,
      query: *mut pg_sys::Query,
      #[cfg(any(feature = "pg14"))]
      jumble_state: *mut JumbleState
  ) {
```
```
❯ cargo build --features pg11 --no-default-features
   Compiling pgx-pg-sys v0.5.3 (/home/eng/pgx/pgx-pg-sys)
   Compiling pgx v0.5.3 (/home/eng/pgx/pgx)
error[E0425]: cannot find value `jumble_state` in this scope
   --> pgx/src/hooks.rs:493:5
    |
493 |     jumble_state: *mut JumbleState
    |     ^^^^^^^^^^^^ not found in this scope
```

Also, I think we'd need 3 cfg blocks to make it work. Something like this:
```rust
+    #[cfg(any(feature = "pg14"))]
+    jumble_state: *mut JumbleState
 ) {
     fn prev(
         parse_state: PgBox<pg_sys::ParseState>,
         query: PgBox<pg_sys::Query>,
-        _jumble_state: Option<PgBox<JumbleState>>,
+        jumble_state: Option<PgBox<JumbleState>>,
     ) -> HookResult<()> {
         HookResult::new(unsafe {
             match HOOKS.as_mut().unwrap().prev_post_parse_analyze_hook.as_ref() {
                 None => (),
+                #[cfg(any(feature = "pg10", feature = "pg11", feature="pg12", feature="pg13"))]
                 Some(f) => (f)(parse_state.as_ptr(), query.as_ptr()),
+                #[cfg(any(feature = "pg14"))]
+                Some(f) => (f)(parse_state.as_ptr(), query.as_ptr(), jumble_state.unwrap().as_ptr()),
```

**Comment by workingjubilee** [comment]
Ooh, the extra parameter part I didn't notice.

**Comment by workingjubilee** [comment]
That might be hard to work around due to https://github.com/tcdi/pgx/issues/794 but hopefully that will be fixed soon.

File: pgx/src/hooks.rs
Comment (workingjubilee) [inline pgx/src/hooks.rs:542]: This would eliminate the duplication, and works even with #794 not being fixed, but considering the macro magic, I am not sure it's worth it. What do you think?

```suggestion
macro_rules! pgx_post_parse_analyze_body {
    ($parse_state: ident, $query: ident$(, $_jumble_state: ident)?) => {
        {
            fn prev(
                parse_state: PgBox<pg_sys::ParseState>,
                query: PgBox<pg_sys::Query>,
                _jumble_state: Option<PgBox<JumbleState>>,
            ) -> HookResult<()> {
                HookResult::new(unsafe {
                    match HOOKS.as_mut().unwrap().prev_post_parse_analyze_hook.as_ref() {
                        None => (),
                        #[cfg(any(feature = "pg10", feature = "pg11", feature = "pg12", feature = "pg13"))]
                        Some(f) => (f)(parse_state.as_ptr(), query.as_ptr()),
                        #[cfg(any(feature = "pg14"))]
                        Some(f) => (f)(parse_state.as_ptr(), query.as_ptr(), _jumble_state.unwrap().as_ptr()),
                    }
                })
            }
            #[cfg(any(feature = "pg10", feature = "pg11", feature = "pg12", feature = "pg13"))]
            let jumble_box = None;
            #[cfg(any(feature = "pg14"))]
            let jumble_box =  Some(PgBox::from_pg($($_jumble_state)?));

            let hook = &mut HOOKS.as_mut().unwrap().current_hook;
            hook.post_parse_analyze(
                PgBox::from_pg($parse_state),
                PgBox::from_pg($query),
                jumble_box,
                prev,
            )
            .inner
        }
    }
}

#[cfg(any(feature = "pg10", feature = "pg11", feature = "pg12", feature = "pg13"))]
#[pg_guard]
unsafe extern "C" fn pgx_post_parse_analyze(
    parse_state: *mut pg_sys::ParseState,
    query: *mut pg_sys::Query,
) {
    pgx_post_parse_analyze_body!(parse_state, query)
}

#[cfg(any(feature = "pg14"))]
#[pg_guard]
unsafe extern "C" fn pgx_post_parse_analyze(
    parse_state: *mut pg_sys::ParseState,
    query: *mut pg_sys::Query,
    _jumble_state: *mut JumbleState,
) {
    pgx_post_parse_analyze_body!(parse_state, query, _jumble_state)
}

```

Comment (jteplitz) [reply pgx/src/hooks.rs:542]: Jubilee and I discussed this on discord and agreed that this is probably more magic than it's worth.

**Comment by workingjubilee** [comment]
https://github.com/tcdi/pgx/pull/813 was merged, you might want to take a second crack at this?

File: pgx-pg-sys/include/pg10.h
Comment (workingjubilee) [inline pgx-pg-sys/include/pg10.h:77]:
```suggestion
```
This is about to go away anyways.

File: pgx-pg-sys/include/pg11.h
Comment (workingjubilee) [inline pgx-pg-sys/include/pg11.h:75]: Are there any C macros from this header that you want to bind?

**Review by workingjubilee — CHANGES_REQUESTED** [review]
You probably will want to also support PostgreSQL 15.

**Comment by workingjubilee** [comment]
I kind of wanted #794 to be fully fixed before landing this, but I don't think it's actually worth it to keep it open. So, I will approve and merge this after #830 lands and this is updated accordingly.

File: pgx-pg-sys/include/pg11.h
Comment (jteplitz) [reply pgx-pg-sys/include/pg11.h:75]: I don't think so. Why do you ask? I just added this to get the definitions of `pgx_post_parse_analyze` and `pgx_post_parse_analyze_type`

Comment (workingjubilee) [reply pgx-pg-sys/include/pg11.h:75]: Just making sure since C macros aren't bound by bindgen.

**Review by workingjubilee — APPROVED** [review]
Thank you!

**Pull request merged**
